private/buf/cmd/buf/command/push/push.go

```diff
@@ -347,6 +388,306 @@ func validateLabelFlagValues(flags *flags) error {
+		}
+		if flags.Branch != "" {
+			usedFlags = append(usedFlags, branchFlagName)
+		}
+		if flags.Draft != "" {
+			usedFlags = append(usedFlags, draftFlagName)
+		}
+		if len(usedFlags) > 0 {
+			usedFlagsErrStr := strings.Join(
+				slicesext.Map(
+					usedFlags,
+					func(flag string) string { return fmt.Sprintf("--%s", flag) },
+				),
+				", ",
+			)
+			return appcmd.NewInvalidArgumentErrorf(
+				"The following flag(s) cannot be used in combination with --%s: %s.",
+				gitMetadataFlagName,
+				usedFlagsErrStr,
+			)
+		}
+	}
+	return nil
+}
+
+func getGitMetadataUploadOptions(
+	ctx context.Context,
+	container appext.Container,
+	flags *flags,
+) ([]bufmodule.UploadOption, error) {
+	input, err := bufcli.GetInputValue(container, flags.InputHashtag, ".")
+	if err != nil {
+		return nil, err
+	}
+	runner := command.NewRunner()
+	var gitMetadataUploadOptions []bufmodule.UploadOption
+	gitLabelsUploadOption, err := getGitMetadataLabelsUploadOptions(ctx, runner, input)
+	if err != nil {
+		return nil, err
+	}
+	if gitLabelsUploadOption != nil {
+		gitMetadataUploadOptions = append(gitMetadataUploadOptions, gitLabelsUploadOption)
+	}
+	remotes, err := getGitRemotes(ctx, runner, input)
+	if err != nil {
+		return nil, err
+	}
+	gitSourceControlURLUploadOption, err := getGitMetadataSourceControlURLUploadOption(ctx, runner, remotes, input)
+	if err != nil {
+		return nil, err
+	}
+	if gitSourceControlURLUploadOption != nil {
+		gitMetadataUploadOptions = append(gitMetadataUploadOptions, gitSourceControlURLUploadOption)
+	}
+	if flags.Create {
+		gitDefaultBranch, err := getGitDefaultBranch(ctx, runner, container, remotes, input)
+		if err != nil {
+			return nil, err
+		}
+		createModuleVisibility, err := bufmodule.ParseModuleVisibility(flags.CreateVisibility)
+		if err != nil {
+			return nil, err
+		}
+		gitMetadataUploadOptions = append(
+			gitMetadataUploadOptions,
+			bufmodule.UploadWithCreateIfNotExist(createModuleVisibility, gitDefaultBranch),
+		)
+	}
+	return gitMetadataUploadOptions, nil
+}
+
+func getGitRemotes(
+	ctx context.Context,
+	runner command.Runner,
+	input string,
+) ([]string, error) {
+	buffer := bytes.NewBuffer(nil)
+	if err := runner.Run(
+		ctx,
+		gitCommand,
+		command.RunWithArgs("remote"),
+		command.RunWithStdout(buffer),
+		command.RunWithStderr(buffer),
+		command.RunWithDir(input),
+	); err != nil {
+		return nil, err
+	}
+	return strings.Split(strings.TrimSpace(buffer.String()), "\n"), nil
+}
+
+// This returns an upload option for all Git metadata labels. We set labels for the following
+// Git matadata:
+//   - tags on current commit
+//   - current branch
+func getGitMetadataLabelsUploadOptions(
+	ctx context.Context,
+	runner command.Runner,
+	input string,
+) (bufmodule.UploadOption, error) {
+	tags, err := getGitTagsOnCurrentCommit(ctx, runner, input)
+	if err != nil {
+		return nil, err
+	}
+	branch, err := getCurrentGitBranch(ctx, runner, input)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Interesting, I would expect this to be the set of all branches that point to this commit, not just the current branch. Wouldn't that get us closer to what is in Git, and also pushes on other branches no-ops due to squashing? Otherwise a single git commit that is on multiple branches will not have a corresponding commit in the BSR with multiple branches.

I can see the value in being context-centric though -- this isn't `sync`, after all. What do we think?

Reply:
Just updating the comment here based on the conversations we've had external to this PR: it would be all branches that point directly to the commit at the time of the command.